File: labodc-web-portal/src/services/admin/dashboardService.ts
// src/services/admin/dashboardService.ts

import axios from 'axios';

// Use relative URL to leverage Vite proxy configuration
const API_BASE_URL = '/api';

// Axios instance với config
const api = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Interceptor để thêm token
api.interceptors.request.use((config) => {
  const token = localStorage.getItem('access_token');
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
    console.log('🔑 Request with token:', token.substring(0, 30) + '...');
  } else {
    console.warn('⚠️ No token found in localStorage!');
  }
  return config;
});

// Response interceptor for debugging
api.interceptors.response.use(
  (response) => {
    console.log('📥 Response received:', {
      url: response.config.url,
      status: response.status,
      data: response.data,
    });
    return response;
  },
  (error) => {
    console.error('❌ API Error:', {
      url: error.config?.url,
      status: error.response?.status,
      data: error.response?.data,
    });
    return Promise.reject(error);
  }
);

// Types theo SRS document
export interface DashboardStats {
  projects: {
    total: number;
    new: number;
    ongoing: number;
    completed: number;
    cancelled: number;
    successRate: number;
  };
  enterprises: {
    total: number;
    new: number;
    active: number;
    verified: number;
  };
  talents: {
    total: number;
    new: number;
    active: number;
    averageRating: number;
  };
  mentors: {
    total: number;
    active: number;
    averageRating: number;
  };
  financials: {
    totalRevenue: number;
    teamDisbursed: number;
    mentorDisbursed: number;
    labRevenue: number;
    hybridFundAdvanced: number;
    hybridFundRepaid: number;
  };
  performance: {
    avgProjectCompletion: number;
    onTimeDelivery: number;
    customerSatisfaction: number;
  };
}

export interface RecentActivity {
  id: number;
  type: 'project' | 'payment' | 'enterprise' | 'report';
  title: string;
  description: string;
  timestamp: string;
  status: 'success' | 'warning' | 'info' | 'error';
}

export interface PendingApproval {
  id: number;
  type: 'enterprise' | 'project';
  title: string;
  submittedAt: string;
  priority: 'high' | 'medium' | 'low';
  
  // Enterprise specific fields
  companyName?: string;
  taxCode?: string;
  businessLicenseNumber?: string;
  address?: string;
  city?: string;
  district?: string;
  ward?: string;
  representativeName?: string;
  representativePosition?: string;
  contactEmail?: string;
  contactPhone?: string;
  website?: string;
  industry?: string;
  companySize?: string;
  yearEstablished?: number;
  description?: string;
  
  // Project specific fields
  slug?: string;
  requirements?: string;
  startDate?: string;
  endDate?: string;
  budget?: number;
  currency?: string;
  numberOfStudents?: number;
  status?: string;
  enterpriseName?: string;
}

// API Service
class DashboardService {
  // Lấy thống kê tổng quan
  async getStats(): Promise<DashboardStats> {
    console.log('📊 Loading dashboard stats from backend...');
    
    const response = await api.get('/dashboard/stats');
    console.log('✅ Full response:', response);
    console.log('✅ response.data:', response.data);
    console.log('✅ response.data.data:', response.data.data);
    
    // Map backend response to frontend format
    const data = response.data.data;
    
    if (!data) {
      console.error('❌ Backend response missing data field!');
      throw new Error('Invalid backend response structure');
    }
    
    return {
      projects: {
        total: data.projects.total,
        new: data.projects.newCount,
        ongoing: data.projects.ongoing,
        completed: data.projects.completed,
        cancelled: data.projects.cancelled,
        successRate: data.projects.successRate,
      },
      enterprises: {
        total: data.enterprises.total,
        new: data.enterprises.newCount,
        active: data.enterprises.active,
        verified: data.enterprises.verified,
      },
      talents: {
        total: data.talents.total,
        new: data.talents.newCount,
        active: data.talents.active,
        averageRating: data.talents.averageRating,
      },
      mentors: {
        total: data.mentors.total,
        active: data.mentors.active,
        averageRating: data.mentors.averageRating,
      },
      financials: {
        totalRevenue: data.financials.totalRevenue,
        teamDisbursed: data.financials.teamDisbursed,
        mentorDisbursed: data.financials.mentorDisbursed,
        labRevenue: data.financials.labRevenue,
        hybridFundAdvanced: data.financials.hybridFundAdvanced,
        hybridFundRepaid: data.financials.hybridFundRepaid,
      },
      performance: {
        avgProjectCompletion: data.performance.avgProjectCompletion,
        onTimeDelivery: data.performance.onTimeDelivery,
        customerSatisfaction: data.performance.customerSatisfaction,
      },
    };
  }

  // Lấy hoạt động gần đây
  async getRecentActivities(): Promise<RecentActivity[]> {
    console.log('📝 Loading recent activities from backend...');
    
    const response = await api.get('/dashboard/activities?limit=10');
    console.log('✅ Activities loaded from backend:', response.data.data.length);
    
    // Map backend response to frontend format
    return response.data.data.map((activity: any) => ({
      id: activity.id,
      type: activity.activityType,
      title: activity.title,
      description: activity.description,
      timestamp: activity.timestamp,
      status: activity.status,
    }));
  }

  // Lấy danh sách chờ phê duyệt
  async getPendingApprovals(limit: number = 10): Promise<PendingApproval[]> {
    console.log('⏳ Loading pending approvals from backend...');
    
    const response = await api.get(`/dashboard/approvals?limit=${limit}`);
    console.log('✅ Approvals loaded from backend:', response.data.data);
    
    // Return full data from backend (already matches interface)
    return response.data.data;
  }

  // Approve enterprise
  async approveEnterprise(id: number): Promise<void> {
    const response = await api.put(`/enterprises/${id}/verify`);
    return response.data;
  }

  // Reject enterprise
  async rejectEnterprise(id: number, reason?: string): Promise<void> {
    const response = await api.delete(`/enterprises/${id}`, {
      data: { reason }
    });
    return response.data;
  }

  // Approve project
  async approveProject(id: number): Promise<void> {
    const response = await api.put(`/projects/${id}/validate`);
    return response.data;
  }

  // Reject project
  async rejectProject(id: number, reason?: string): Promise<void> {
    const response = await api.delete(`/projects/${id}`, {
      data: { reason }
    });
    return response.data;
  }

  // Lấy dữ liệu biểu đồ doanh thu theo tháng
  async getRevenueChart(months: number = 6): Promise<any[]> {
    console.log('📈 Loading revenue chart from backend...');
    
    const response = await api.get(`/dashboard/revenue?months=${months}`);
    console.log('✅ Chart data loaded from backend:', response.data.data.length);
    
    // Map backend response to frontend format
    return response.data.data.map((item: any) => ({
      month: item.month,
      revenue: item.revenue,
      teamDisbursed: item.teamDisbursed,
      mentorDisbursed: item.mentorDisbursed,
      labRevenue: item.labRevenue,
    }));
  }
}

export const dashboardService = new DashboardService();

File: labodc_mobile/lib/services/admin/dashboard_service.dart
import 'package:dio/dio.dart';
import 'package:labodc_mobile/services/api_service.dart';
import 'package:logger/logger.dart';

/// Dashboard Service for Lab Admin
class DashboardService {
  final ApiService _api = ApiService();
  final _logger = Logger();

  /// Get dashboard statistics
  Future<DashboardStats> getStats() async {
    try {
      _logger.i('📊 Loading dashboard stats...');
      final response = await _api.get('/dashboard/stats');
      
      final data = response.data['data'];
      _logger.i('✅ Dashboard stats loaded');
      
      return DashboardStats.fromJson(data);
    } catch (e) {
      _logger.e('❌ Error loading dashboard stats', error: e);
      throw Exception('Không thể kết nối đến server. Vui lòng kiểm tra:\n- Backend có đang chạy ở http://localhost:8080\n- Kết nối mạng của bạn');
    }
  }

  /// Get recent activities
  Future<List<RecentActivity>> getRecentActivities() async {
    try {
      _logger.i('📋 Loading recent activities...');
      final response = await _api.get('/dashboard/activities');
      
      final data = response.data['data'] as List;
      _logger.i('✅ Loaded ${data.length} activities');
      
      return data.map((json) => RecentActivity.fromJson(json)).toList();
    } catch (e) {
      _logger.e('❌ Error loading activities', error: e);
      rethrow;
    }
  }

  /// Get pending approvals
  Future<List<PendingApproval>> getPendingApprovals(int limit) async {
    try {
      _logger.i('⏳ Loading pending approvals...');
      final response = await _api.get('/dashboard/approvals', queryParameters: {'limit': limit});
      
      final data = response.data['data'] as List;
      _logger.i('✅ Loaded ${data.length} pending approvals');
      
      return data.map((json) => PendingApproval.fromJson(json)).toList();
    } catch (e) {
      _logger.e('❌ Error loading pending approvals', error: e);
      rethrow;
    }
  }

  /// Get revenue chart data
  Future<List<RevenueChart>> getRevenueChart(int months) async {
    try {
      _logger.i('📈 Loading revenue chart...');
      final response = await _api.get('/dashboard/revenue', queryParameters: {'months': months});
      
      final data = response.data['data'] as List;
      _logger.i('✅ Loaded revenue data for $months months');
      
      return data.map((json) => RevenueChart.fromJson(json)).toList();
    } catch (e) {
      _logger.e('❌ Error loading revenue chart', error: e);
      rethrow;
    }
  }

  /// Approve enterprise
  Future<void> approveEnterprise(int id) async {
    try {
      await _api.put('/enterprises/$id/verify');
      _logger.i('✅ Enterprise $id approved');
    } catch (e) {
      _logger.e('❌ Error approving enterprise', error: e);
      rethrow;
    }
  }

  /// Reject enterprise
  Future<void> rejectEnterprise(int id, String? reason) async {
    try {
      await _api.delete('/enterprises/$id', data: {'reason': reason});
      _logger.i('✅ Enterprise $id rejected');
    } catch (e) {
      _logger.e('❌ Error rejecting enterprise', error: e);
      rethrow;
    }
  }

  /// Approve project
  Future<void> approveProject(int id) async {
    try {
      await _api.put('/projects/$id/validate');
      _logger.i('✅ Project $id approved');
    } catch (e) {
      _logger.e('❌ Error approving project', error: e);
      rethrow;
    }
  }

  /// Reject project
  Future<void> rejectProject(int id, String? reason) async {
    try {
      await _api.put('/projects/$id/reject', data: {'reason': reason});
      _logger.i('✅ Project $id rejected');
    } catch (e) {
      _logger.e('❌ Error rejecting project', error: e);
      rethrow;
    }
  }
}

// ============= MODELS =============

class DashboardStats {
  final ProjectStats projects;
  final EnterpriseStats enterprises;
  final TalentStats talents;
  final MentorStats mentors;
  final double totalRevenue;

  DashboardStats({
    required this.projects,
    required this.enterprises,
    required this.talents,
    required this.mentors,
    required this.totalRevenue,
  });

  factory DashboardStats.fromJson(Map<String, dynamic> json) {
    return DashboardStats(
      projects: ProjectStats.fromJson(json['projects']),
      enterprises: EnterpriseStats.fromJson(json['enterprises']),
      talents: TalentStats.fromJson(json['talents']),
      mentors: MentorStats.fromJson(json['mentors']),
      totalRevenue: (json['totalRevenue'] ?? 0).toDouble(),
    );
  }
}

class ProjectStats {
  final int total;
  final int newCount;
  final int ongoing;
  final int completed;
  final int cancelled;
  final double successRate;

  ProjectStats({
    required this.total,
    required this.newCount,
    required this.ongoing,
    required this.completed,
    required this.cancelled,
    required this.successRate,
  });

  factory ProjectStats.fromJson(Map<String, dynamic> json) {
    return ProjectStats(
      total: json['total'] ?? 0,
      newCount: json['newCount'] ?? 0,
      ongoing: json['ongoing'] ?? 0,
      completed: json['completed'] ?? 0,
      cancelled: json['cancelled'] ?? 0,
      successRate: (json['successRate'] ?? 0).toDouble(),
    );
  }
}

class EnterpriseStats {
  final int total;
  final int pending;
  final int verified;
  final int rejected;

  EnterpriseStats({
    required this.total,
    required this.pending,
    required this.verified,
    required this.rejected,
  });

  factory EnterpriseStats.fromJson(Map<String, dynamic> json) {
    return EnterpriseStats(
      total: json['total'] ?? 0,
      pending: json['pending'] ?? 0,
      verified: json['verified'] ?? 0,
      rejected: json['rejected'] ?? 0,
    );
  }
}

class TalentStats {
  final int total;
  final int active;
  final int available;
  final int working;

  TalentStats({
    required this.total,
    required this.active,
    required this.available,
    required this.working,
  });

  factory TalentStats.fromJson(Map<String, dynamic> json) {
    return TalentStats(
      total: json['total'] ?? 0,
      active: json['active'] ?? 0,
      available: json['available'] ?? 0,
      working: json['working'] ?? 0,
    );
  }
}

class MentorStats {
  final int total;
  final int active;
  final int available;
  final int mentoring;

  MentorStats({
    required this.total,
    required this.active,
    required this.available,
    required this.mentoring,
  });

  factory MentorStats.fromJson(Map<String, dynamic> json) {
    return MentorStats(
      total: json['total'] ?? 0,
      active: json['active'] ?? 0,
      available: json['available'] ?? 0,
      mentoring: json['mentoring'] ?? 0,
    );
  }
}

class RecentActivity {
  final int id;
  final String type;
  final String action;
  final String description;
  final String? userName;
  final DateTime timestamp;

  RecentActivity({
    required this.id,
    required this.type,
    required this.action,
    required this.description,
    this.userName,
    required this.timestamp,
  });

  factory RecentActivity.fromJson(Map<String, dynamic> json) {
    return RecentActivity(
      id: json['id'] ?? 0,
      type: json['type'] ?? '',
      action: json['action'] ?? '',
      description: json['description'] ?? '',
      userName: json['userName'],
      timestamp: DateTime.parse(json['timestamp']),
    );
  }
}

class PendingApproval {
  final int id;
  final String type; // ENTERPRISE, PROJECT
  final String name;
  final String? submittedBy;
  final DateTime submittedAt;
  final String? priority; // HIGH, NORMAL, LOW
  final String? status;

  PendingApproval({
    required this.id,
    required this.type,
    required this.name,
    this.submittedBy,
    required this.submittedAt,
    this.priority,
    this.status,
  });

  factory PendingApproval.fromJson(Map<String, dynamic> json) {
    return PendingApproval(
      id: json['id'] ?? 0,
      type: json['type'] ?? '',
      name: json['name'] ?? '',
      submittedBy: json['submittedBy'],
      submittedAt: DateTime.parse(json['submittedAt']),
      priority: json['priority'],
      status: json['status'],
    );
  }
}

class RevenueChart {
  final String month;
  final double revenue;
  final double expenses;
  final double profit;

  RevenueChart({
    required this.month,
    required this.revenue,
    required this.expenses,
    required this.profit,
  });

  factory RevenueChart.fromJson(Map<String, dynamic> json) {
    return RevenueChart(
      month: json['month'] ?? '',
      revenue: (json['revenue'] ?? 0).toDouble(),
      expenses: (json['expenses'] ?? 0).toDouble(),
      profit: (json['profit'] ?? 0).toDouble(),
    );
  }
}
